core/shared/src/main/scala/org/http4s/multipart/MultipartReceiver.scala

```diff
@@ -0,0 +1,192 @@
+      def assemble(partials: List[A]): Either[DecodeFailure, Option[A]] = partials match {
+        case Nil => Right(None)
+        case one :: Nil => Right(Some(one))
+        case _ => Left(fail(s"Unexpected multiple parts found with name '$name'"))
+      }
+    }
+
+    def asList: MultipartReceiver[F, List[A]] = new ReceiverAt(name, receiver)
+      with MultipartReceiver[F, List[A]] {
+      def assemble(partials: List[Partial]): Either[DecodeFailure, List[A]] = Right(partials)
+    }
+  }
+
+  private abstract class ReceiverAt[F[_], A](name: String, receiver: PartReceiver[F, A]) {
+    type Partial = A
+    def decide(headers: Headers): Option[PartReceiver[F, A]] =
+      if (partName(headers).contains(name)) Some(receiver)
+      else None
+  }
+
+  def auto[F[_]: Concurrent: Files]: MultipartReceiver[F, Map[String, PartValue]] =
+    new MultipartReceiver[F, Map[String, PartValue]] {
+      type Partial = (String, PartValue)
+      def decide(partHeaders: Headers): Option[PartReceiver[F, Partial]] =
+        (partName(partHeaders), partFilename(partHeaders)) match {
+          case (Some(name), Some(filename)) =>
+            Some(PartReceiver.toTempFile[F].map { path =>
+              name -> PartValue.OfFile(filename, path)
+            })
+          case (Some(name), None) =>
+            Some(PartReceiver.bodyText[F].map(str => name -> PartValue.OfString(str)))
+          case _ => None
+        }
+      def assemble(partials: List[Partial]): Either[DecodeFailure, Map[String, PartValue]] = Right(
+        partials.toMap
+      )
+    }
+
+  def uniform[F[_], A](partReceiver: PartReceiver[F, A]): MultipartReceiver.Aux[F, List[A], A] =
+    new MultipartReceiver[F, List[A]] {
+      type Partial = A
+      def decide(partHeaders: Headers) = Some(partReceiver)
+      def assemble(partials: List[Partial]) = Right(partials)
+    }
+
+  private class IgnoreUnexpected[F[_], A, P](inner: Aux[F, A, P]) extends MultipartReceiver[F, A] {
+    type Partial = Option[P]
+    def decide(partHeaders: Headers): Option[PartReceiver[F, Partial]] =
+      Some(inner.decide(partHeaders).map(_.map[Partial](Some(_))).getOrElse {
+        PartReceiver.ignore[F].map[Partial](_ => None)
+      })
+    def assemble(partials: List[Partial]): Either[DecodeFailure, A] =
+      inner.assemble(partials.flatten)
+  }
+
+  private class TransformPart[F[_], A, P](
+      inner: Aux[F, A, P],
+      f: PartReceiver[F, P] => PartReceiver[F, P],
+  ) extends MultipartReceiver[F, A] {
+    type Partial = P
+    def decide(partHeaders: Headers): Option[PartReceiver[F, P]] = inner.decide(partHeaders).map(f)
+    def assemble(partials: List[P]): Either[DecodeFailure, A] = inner.assemble(partials)
+  }
+
+  implicit def multipartReceiverApplicative[F[_]]: Applicative[MultipartReceiver[F, *]] =
+    new MultipartReceiverApplicative[F]
+
+  private class MultipartReceiverApplicative[F[_]] extends Applicative[MultipartReceiver[F, *]] {
```

Comment:
An instance of an unsealed type makes me a little nervous.  I'm getting too sleepy to think about whether the abstract methods could affect the legality of the instance.

Reply:
I'm not sure I understand what you're saying here. Is the issue with defining the `private class` that extends `Applicative`, as opposed to just returning an anonymous class from `def multipartReceiverApplicative`? Or is this more related to "cats laws" legality? 

Assuming it's about cats laws, I initially inquired about this in the typelevel discord's #cats channel - [here](https://discord.com/channels/632277896739946517/632278570512678923/1217670243896262756). I got the impression that the "order matters" nature of `decide` when combining `MultipartReceiver`s was acceptable.